### ISSUE TITLE
[Feature Request] "Classic Flyers" mod statistics adjustment

### ISSUE THREAD

**Issue by Storm-in-Crimson** [cadon/ARKStatsExtractor#244]
Due to update v. 256, a lot of players (including our whole server) switched to former version of flyers mechanics, using "**Classic Flyers**" mod. There are over 60.000 people subscribed on mod's page, so it will be really nice to implement new program option for them. As for now your awesome program can't read stats of flyers, as they're swapped in "Classic Flyers". I think it will be awesome to add such an option :). Is it possible or maybe planned?

**Comment by cadon** [comment]
It's already kind of implemented. You only need a file with the content of
`{"species":[{"name":"Pteranodon","statsRaw":[[200,0.1],null,null,null,[1000,0.08,0.9,0,0]]},{"name":"Quetzal","statsRaw":[[200,0.1],null,null,null,[1000,0.08,0.9,0,0]]}]}`
(these values are not the correct ones for the mod, just as an example), and load them via File -> Load additional values. All values that are `null` or left, stay the original ones from the official servers. Order of the stats is: Base, wildLevel, tamedLevel, tamingAdd, tamingAffinity. The stat's order is HP, St, Ox, Fo, We, Dm, Sp, To.
I couldn't find the values the mod uses with a quick search, if you have a list of them, I can help to create the needed file.

**Comment by Storm-in-Crimson** [comment]
Great there's already what's basically needed to make swift transition :) 
I'm not sure if this are the stats required for making custom json-file for Classic Flyers, but please check it out:
http://steamcommunity.com/workshop/filedetails/discussion/895711211/133261907130730127/

If its not all or any of the values needed, I'll try to reach mod's author and ask him to help provide the full list needed (in some species cases it'll be more than only Stamina & Speed, I guess, like Quetzal's). Basically we need all the values for the stats this mod is changing in regard of flyers, right? 

Values of: Base, wildLevel, tamedLevel, tamingAdd, tamingAffinity - seems not affected by this mod, right?

Thanks for your offer to help, really appreciate it. Was using your program a lot and happily advertised it among fellow Ark players :) Sorry for my dim orientation about the numbers and statistics required to make transition for Classic Flyers, I'm not great when it comes to numbers, will need directions in "for dummies" kind of way ;).

**Comment by cadon** [comment]
These values look good. If that's all the values that the mod changes, the file I just uploaded should work with that. You can download the file here: https://raw.githubusercontent.com/cadon/ARKStatsExtractor/master/ARKBreedingStats/classicalFlyers.json, and load it into the tool via File -> Load additional values.
Let me know if that works.

**Comment by Storm-in-Crimson** [comment]
Ok, I'm pestering Ertosi (Classic Flyers author) if Stamina & Speed are
only stats changed, and Weight or HP (or anything other) are left vanilla.
Just to make sure we have all the data. Will give you feedback after he
responds.

I'm so grateful you found the solution, was seriously worried I won't be
able to use Smart Breeding anymore, after switching to "Classic Flyers".
And I really appreciate your app :) Its a "must-have" for controlling new
tames stats and making long-term breeding plans :)

Thanks!

2017-04-09 17:51 GMT+02:00 cadon <notifications@github.com>:

**Comment by Storm-in-Crimson** [comment]
Ertosi stated, that only modifications include Stamina & Speed, and all the mod is changing is a roll-back to flyers stats from a couple of months back:

"_All I did was "rollback" the flyer files to a couple of months ago and the table above shows the exact numbers in the files._
(...)
_All of their stats in this mod are exactly what they used to be pre-nerf. Their Speed and Stamina are different in-game not due to base stats, but because post-nerf the formulas in the core game to calculate the final speed and stamina from these base numbers are now different (how so, we can't see or access). Hence why we've come here - to decide the best way to get their perceived stats back to their old feel._"

So I got one additional question - maybe with your data for Smart Breeding we can help Ertosi to calibrate flyers speed? Will it be possible?

**Comment by cadon** [comment]
As far as I know only the base-values and the increase-per-level (wild and tamed) were nerfed, by reverting these three numbers for each stamina and speed it should be as before. The increase per wild and tamed level for stamina was 10% (=0.1), the speed-increase and the base-values are listed in the table as well, so I guess the file I linked should work. As I see it, there are no more numbers to be considered to revert the nerf.

Did you try the additional-values, did it work?

**Comment by smw** [comment]
It doesn't appear to be working for me, at least for Quetzals or Pteranodons.  I'm often seeing health (?!) and speed as red.   I really appreciate your work on this, sent you a donation to support!

**Comment by cadon** [comment]
Hi @smw, thanks a lot!
The HP were nerfed recently with a new introduced multiplier, probably that multiplier is also reverted by the mod, I'll expand the functionality to revert this in the tool as well.
For the speed I can only explain it if the creatures were imprinted, there was another change.

**Comment by Storm-in-Crimson** [comment]
I checked main flyers species (the Island) stats extraction with updated classicalFlyers.json, and the results shows problems (invalid extraction, red boxes) in:

- Argentavis - Weight
- Pelagornis - Health
- Pteranodon - Health & Weight
- Quetzal - Health & Weight
- Tapejara - Weight

Also, in every case, extraction shows Speed stat as 0 lvl-ups. Not sure, but if Ertosi only rolled back, it should be above 0.
Dinos were freshly spawned for the testing purposes ('GmSummons' command), with active Classic Flyers mod.

**Comment by cadon** [comment]
Then the additional values still seem to miss some of the original values, or they were changed by the mod. As I understood it, the mod-author does some tweaks to get a good balance. You can give it a try, I wrote a short tutorial how to create these files here https://github.com/cadon/ARKStatsExtractor/wiki/Mod-Values, a good tool to edit json-files is on http://www.jsoneditoronline.org/.
I'm on vacation for the next two weeks, so I'll not be able to give support for some time. Good luck :)

**Comment by Wardjp** [comment]
Same here, getting red on weight when attempting to export. As I have time I'll try messing with the file creations myself and let you know my results. Thanks for all the hard work you do and enjoy your vacation!

**Comment by Storm-in-Crimson** [comment]
Have a nice vacation, we'll sort this out eventually, after Easter's full swing :)

**Comment by TATTOOGUY** [comment]
Hey guys, I'm still having this issue. Has anyone come across a working fix for the problems listed by Storm-in-Crimson above?

    Argentavis - Weight
    Pelagornis - Health
    Pteranodon - Health & Weight
    Quetzal - Health & Weight
    Tapejara - Weight

**Comment by TATTOOGUY** [comment]
...anyone?

**Comment by Storm-in-Crimson** [comment]
Nothing here, I'm afraid. 
While Cadon left instruction how to adjust it, as I understand it - its a matter of testing the formula (http://ark.gamepedia.com/Creature_Stats_Calculation) with changing variables values (in this case I suspect "_increase per wild-level as % of B: Iw_" - Cadon's **wildLevel** and "_increase per domesticated level as % of B: Id_" - **tamedLevel**), till we get it right and stats in game will match those in SmartBreeding app. 

Not sure, but Base value might also been changed for flyers species.

Then, there's also TamedBaseHealthMultiplier added in update .256 to consider - it messes with some species Health. Guess it might be also responsible for improper Health stat extractions.

According to Cadon's description, in the example below, Argent's "additional values" in ClassicFlyers.json are:
{"species":[ {"name":"Argentavis","statsRaw":[null,[750,0.1,0.1],null,null,null,null,[null,null,0.025],null]},

I understand that means that Argent's Health is not changed for this mod (its value is "null"), then Stamina is changed (Base value = 750, wildLevel=0.1, tamedLevel=0.1). Next 4 stats Ox, Fo, We, Dm - are left unchanged. Speed is changed in either wildLevel or tamedLevel (not sure which one it is), overriding general value with 0.025 for mod's users. The last stat, Torpor - is also left as "null", so not changed.
Not sure where goes tamingAdd and tamingAffinity - is it safe to assume it was not changed as well for flyers, or if we need to check this too.

So, a lot of calculations - I'm clueless where to begin :/

**Comment by smw** [comment]
I've been able to get most flyers to load by mucking with the weight stat.
  Usually I set it to a round number, hit up or down a few clicks and hit
enter after each one.

**Comment by Storm-in-Crimson** [comment]
Huh, dunno if this is 100% correct, but after being stuck in dead end, I tried different approach. 
Loaded **values.json** from **SmartBreeding v. 0.22.1** (just before the fatal .256 Ark update), and searched for Argent's Weight values there. Then "copy-pasted" it to classicalFlyers.json in proper place, replacing Cadon's "null" value. Initially got problem with Oxygen stuck on some default value, but after adding to classicalFlyers.json additional command ("doesNotUseOxygen":false) - it seems app extracted my test Argy correctly :)

Check it out, my **classicalFlyers.json** Argy entry now looks like that:
{"name":"Argentavis","statsRaw":[null,[750,0.1,0.1],null,null,[350,0.02,0.04,0,0],null,[null,null,0.025],null],"doesNotUseOxygen":false,"NoImprintingForSpeed":false}

If you change Argy entry as above, let me know if it works ok and extracts your Argy's properly. 

If so - then its only a matter of finding in values.json from v. 0.22.1 all the flyers values that are showing red now, and replacing them.

**Comment by Wardjp** [comment]
Just input one with no levelups, all greens and looks like it output correctly, 98.9 (think my tribemate used raw prime on 5x) and WLvl 150. I'll let you know if I get any inconsistencies.

**Comment by Storm-in-Crimson** [comment]
Well, if extraction turns up green, we're good. In my case it required a bit of switching on/off option "Since Taming no Server Restart", but worked. 
Crude piece of patching up, hope Cadon will refine it when comes back from vacation :)

Here is experimental entry for Pelagornis classicalFlyers.json:
{"name":"Pelagornis","statsRaw":[[210,0.2,0.27,0.5,0],[320,0.1,0.1],null,null,[135,0.02,0.04,0,0],null,[null,null,0.025],null],"doesNotUseOxygen":false,"TamedBaseHealthMultiplier":1,"NoImprintingForSpeed":false}

**Comment by TATTOOGUY** [comment]
With very little adjustment to weight on Argent, this seems to be working for them. Haven't tried pelagornis yet, but have you any info on Pteras? I need to enter about 30 of them... ugh..

**Comment by Storm-in-Crimson** [comment]
This one is for Pteras, but haven't checked it yet. Was done same way as Argy's & Pelagornis, so might work (or not...).

{"name":"Pteranodon","statsRaw":[[210,0.2,0.27,0.5,0],[300,0.1,0.1],null,null,[150,0.02,0.04,0,0],null,[null,null,0.02],null],"doesNotUseOxygen":false,"TamedBaseHealthMultiplier":1,"NoImprintingForSpeed":false}

**Comment by TATTOOGUY** [comment]
Yea, this ptera change was a big fail.... same issues as before. I can't figure it out....

**Comment by JezaGaia** [comment]
I tried the change for argies and I have an issue with HP unfortunately :(
It's an old tame so maybe that's why ? i'll try to tame a new one and see what happens.

**Comment by blazejohnny** [comment]
I thought i was the only one with the problem but you guys got further than me. I think the additive bonus got changed which is giving us errors further down the line.

**Comment by JezaGaia** [comment]
I had tamed all my flyers after installing the mod but they had been spawned before the mod was installed and it seems that's what the problem.

I changed the pteras numbers in the json file, wiped all wild dinos, found a ptera to tame and this one is working fine.
Will make the test with an argie and let you all know but I think we're on the right track but flyers need to be spawned with the mod active otherwise even if the stats can be updated they're not really right and it shows when we try to check them with this tool. 
Well that and the classicflyers.json needs some work but you get what I mean

**Comment by Storm-in-Crimson** [comment]
Numbers for Quetz are:
{"name":"Quetzal","statsRaw":[[1200,0.2,0.27,0.5,0],[750,0.1,0.1],null,null,[780,0.02,0.04,0,0],null,[null,null,0.025],null],"doesNotUseOxygen":false,"TamedBaseHealthMultiplier":1,"NoImprintingForSpeed":false}

Glad it seems working somehow ok, at least it gives us a start to properly assesing flyers stats again. 
Could somebody test and post which stats are problematic in extraction for Scorched Earth flyers species?

**Comment by JezaGaia** [comment]
So I downloaded the new version and it works perfectly with quetzes, pteras and argies but I cannot make it work for pelargonis
I tried the numbers someone posted above too just in case but no luck,  health and weight are red no matter what I do even when checking stats when wild (after ko before it eats anything)

Am I doing it wrong ?

**Comment by cadon** [comment]
Hi, it could be that the mod was changed again. I'm not using that mod, maybe someone here with that mod can confirm / correct the numbers for that.

**Comment by JezaGaia** [comment]
Well I know that the mod hasn't updated since april 27th so I doubt that's the reason but if i'm wrong someone correct me.

**Comment by Wardjp** [comment]
@cadon How would we go about dissecting the mod to figure those values out?

**Comment by Storm-in-Crimson** [comment]
@JezaGaia Dunno if it might be the case, but have you tried checking &
unchecking "Since Taming no Server Restart" option? In case of my tested
Pelagornis, it was the main factor determining successful extraction,
otherwise got red stats also.

2017-05-04 8:31 GMT+02:00 Wardjp <notifications@github.com>:

**Comment by cadon** [comment]
@Wardjp the easiest way would be if the author tells the values, the other way is to knock out / check wild and tamed creatures with different levels. Then with this formula, the different numbers can be extracted: https://ark.gamepedia.com/Creature_Stats_Calculation

**Comment by JezaGaia** [comment]
@ Storm-in-Crimson sorry missed your answer. I tried that and it didn't help. However something strange happened and now all fliers work except argentavis and it's HP that is red.
Even stranger I have an argie that is bred and it's parent is in the extractor with same HP than it's grandfather who is in the extractor but even so the grandchild cannot be input.

The most frustrating thing is that I had them all in but I made a mistake making room for the last big ark patch and deleted the last version of the extractor thinking I was deleting the old one, so I had an old file where the last 4 argies were missing (2 wild and 2 bred) and when I tried to input them it didn't work anymore even though as I said some had almost the same stats than one parent that is in and also that same argie had been extracted without an problem before my mistake.

Maybe I messed up the argie stats by modifying the .json trying to get it to work for the other fliers, could someone tell me if this is correct : 
{"name":"Argentavis","statsRaw":[null,[750,0.1,0.1],null,null,[350,0.02,0.04,0,0],null,[null,null,0.025],null],"doesNotUseOxygen":false,"NoImprintingForSpeed":false}

**Comment by Storm-in-Crimson** [comment]
Try Cadon's new classicalFlyers.json (attached to newest SmartBreeding v.
0.23).
If it won't help, try lvling up Argy in question - just one level, in case
it is showing not up to date statistics. Sometimes lvling up solves the
problem.
Check also if it's not the case of mod malfunction, and if the Argy is able
to lvl up Speed at all.

If all above fail, use Stat Tester and estimate HP lvl ups basing on
parents lvls; till Total Levels on the bottom will show the correct Argy's
lvl. Don't worry if HP value won't be correct, as long as birds level and
lvl ups are right. Stat Testing allows you to include pedigree also, so no
data will be lost.

Your .json seems correct, Cadon's newest one looks like this for Argentavis:
{"name":"Argentavis","statsRaw":[null,[750,0.1,0.1],null,null,[350,0.02,0.04,0,0],null,[null,null,0.025],null],"NoImprintingForSpeed":false}

2017-05-09 15:35 GMT+02:00 JezaGaia <notifications@github.com>:

**Comment by JezaGaia** [comment]
@ Storm-in-Crimson
Thank you for your help, I managed to extract the bred argie, I used the grandparents values and managed to guess what the 2 stats that were added down the line were as well as the level the bird was when born.

It can be leveled, speed included, without a problem, I had already leveled up HP and did again but nothing helped to make the extract function work automatically.

What I cannot manage to do is extract the 2 wild argies that I included in the bloodline to bring up the last 2 stats. The HP doesn't want to input even though when I tamed them I had managed to extract them.
I suppose something in the leveling up isn't working because wild, just tamed or bred with or without imprinting etc are ok but if I take one already leveled up the tool doesn't recognize the HP anymore .
All the other stats are guessed correctly, that's how I managed to know how the breeding values where to "reconstruct" the stats of the last argie of the line. 
The settings are correct compared to my game.ini, I reloaded the file anyway just to be sure but nothing.

Anyway thanks again for your help, I've got the most important bird in the tool it's what matters :)

I'll have to be more careful not to lose my database in the future, and I should also try to avoid leveling up dinos I'll just use for breeding, just in case.

**Comment by Aarondorn2** [comment]
@JezaGaia did you ever get it to work with your original Argies?  I'm having the same exact issue (my first time trying this tool). Everything works perfect except the health, which is red.  I did put post-tame levels in health before using the tool, so maybe that is what is messing it up. I've tried everything from leveling up a different stat, checking all the different boxes, restarting the server, etc.
